src/Sarif.Viewer.VisualStudio/Models/SarifErrorListItem.cs

```diff
@@ -597,112 +602,100 @@ internal void RemapFilePath(string originalPath, string remappedPath)
             _windowFrame = windowFrame;
 
             AttachToDocument();
+
+            return true;
         }
 
         /// <summary>
         /// Attaches to the document using cached properties.
         /// </summary>
-        internal void AttachToDocument()
+        private void AttachToDocument()
         {
             ThreadHelper.ThrowIfNotOnUIThread();
-            LineMarker?.AttachToDocument(_documentName, _documentCookie, _windowFrame);
 
-            foreach (LocationModel location in Locations)
-            {
-                location.LineMarker?.AttachToDocument(_documentName, _documentCookie, _windowFrame);
+            IComponentModel componentModel = (IComponentModel)AsyncPackage.GetGlobalService(typeof(SComponentModel));
+            if (componentModel == null)
+            { 
+                return;
             }
 
-            foreach (LocationModel location in RelatedLocations)
-            {
-                location.LineMarker?.AttachToDocument(_documentName, _documentCookie, _windowFrame);
-            }
+            ISarifLocationProviderFactory sarifLocationProviderFactory = componentModel.GetService<ISarifLocationProviderFactory>();
 
-            foreach (CallTree callTree in CallTrees)
+            // Get a SimpleTagger over the buffer to color
+            IVsTextView vsTextView = SdkUIUtilities.GetTextViewFromFrame(_windowFrame);
+            IWpfTextView wpfTextView = SdkUIUtilities.GetWpfTextView(vsTextView);
+            SimpleTagger<TextMarkerTag> tagger = sarifLocationProviderFactory.GetTextMarkerTagger(wpfTextView.TextBuffer);
+
+            using (tagger.Update())
             {
-                Stack<CallTreeNode> nodesToProcess = new Stack<CallTreeNode>();
+                LineMarker?.TryAttachToDocument(_documentName, _windowFrame);
 
-                foreach (CallTreeNode topLevelNode in callTree.TopLevelNodes)
+                foreach (LocationModel location in Locations)
                 {
-                    nodesToProcess.Push(topLevelNode);
+                    location.LineMarker?.TryAttachToDocument(_documentName, _windowFrame);
                 }
 
-                while (nodesToProcess.Count > 0)
+                foreach (LocationModel location in RelatedLocations)
                 {
-                    CallTreeNode current = nodesToProcess.Pop();
-
-                    if (current.LineMarker?.CanAttachToDocument(_documentName, _documentCookie, _windowFrame) == true)
-                    {
-                        current.LineMarker?.AttachToDocument(_documentName, (long)_documentCookie, _windowFrame);
-                        current.ApplyDefaultSourceFileHighlighting();
-                    }
-
-                    foreach (CallTreeNode childNode in current.Children)
-                    {
-                        nodesToProcess.Push(childNode);
-                    }
+                    location.LineMarker?.TryAttachToDocument(_documentName, _windowFrame);
```

Comment:
Same?

Reply:
Same comments as above. You can have multiple locations for the same document.

---
In reply to: [479670567](https://github.com/microsoft/sarif-visualstudio-extension/pull/186#discussion_r479670567) [](ancestors = 479670567)